python/instrumentation/openinference-instrumentation-llama-index/tests/openinference/instrumentation/llama_index/test_callback.py

```diff
@@ -0,0 +1,354 @@
+from llama_index import ListIndex, ServiceContext
+from llama_index.callbacks import CallbackManager
+from llama_index.llms import OpenAI
+from llama_index.schema import Document, TextNode
+from openinference.instrumentation.llama_index import LlamaIndexInstrumentor
+from openinference.instrumentation.openai import OpenAIInstrumentor
+from openinference.semconv.trace import (
+    DocumentAttributes,
+    EmbeddingAttributes,
+    MessageAttributes,
+    OpenInferenceSpanKindValues,
+    SpanAttributes,
+    ToolCallAttributes,
+)
+from opentelemetry import trace as trace_api
+from opentelemetry.sdk import trace as trace_sdk
+from opentelemetry.sdk.resources import Resource
+from opentelemetry.sdk.trace.export import SimpleSpanProcessor
+from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
+from respx import MockRouter
+
+for name, logger in logging.root.manager.loggerDict.items():
+    if name.startswith("openinference.") and isinstance(logger, logging.Logger):
+        logger.setLevel(logging.DEBUG)
+        logger.handlers.clear()
+        logger.addHandler(logging.StreamHandler())
+
+
+@pytest.mark.parametrize(
+    "is_stream,is_async",
+    [
+        (False, False),
+        (False, True),
+        (True, False),
+        # FIXME: stream + async is not supported by LlamaIndex as of v0.9.33
+        # (True, True),
+    ],
+)
+@pytest.mark.parametrize("status_code", [200, 400])
+def test_callback_llm(
+    is_async: bool,
+    is_stream: bool,
+    status_code: int,
+    respx_mock: MockRouter,
+    in_memory_span_exporter: InMemorySpanExporter,
+    nodes: List[Document],
+    chat_completion_mock_stream: Tuple[List[bytes], List[Dict[str, Any]]],
+) -> None:
+    question = randstr()
+    answer = chat_completion_mock_stream[1][0]["content"] if is_stream else randstr()
+    callback_manager = CallbackManager()
+    query_engine = ListIndex(nodes).as_query_engine(
+        use_async=is_async,
+        streaming=is_stream,
+        callback_manager=callback_manager,
+        service_context=ServiceContext.from_defaults(
+            llm=OpenAI(),
+            callback_manager=callback_manager,
+        ),
+    )
+    respx_kwargs = (
+        {
+            "stream": MockAsyncByteStream(chat_completion_mock_stream[0])
+            if is_async
+            else MockSyncByteStream(chat_completion_mock_stream[0])
+        }
+        if is_stream
+        else {
+            "json": {
+                "choices": [
+                    {
+                        "index": 0,
+                        "message": {"role": "assistant", "content": answer},
+                        "finish_reason": "stop",
+                    }
+                ],
+            }
+        }
+    )
+    url = "https://api.openai.com/v1/chat/completions"
+    respx_mock.post(url).mock(return_value=Response(status_code=status_code, **respx_kwargs))
+
+    with suppress(openai.BadRequestError):
```

Comment:
Is this test case meant to test sending a bad request to the OpenAI? I am getting confused by the checks for different status codes below.

Reply:
yes, this raises an exception when status code is 400 in the mock, but we'll still get some of the spans exported